modules/OFConnectionManager2/module/src/cxn_instance.c

```diff
@@ -0,0 +1,2122 @@
+                              &flag, sizeof(int));
+        }
+
+        LOG_VERBOSE(cxn, "Created non-blocking socket %d, controller %p",
+                    cxn->sd, cxn->controller);
+    }
+
+    cxn->active = true;
+
+    return cxn;
+
+ error:
+    if (cxn && cxn->sd >= 0) {
+        close(cxn->sd);
+        cxn->sd = -1;
+    }
+    return NULL;
+}
+
+
+/**
+ * Disconnect and clean up
+ * @param cxn Connection to be freed
+ */
+void
+ind_cxn_free(connection_t *cxn)
+{
+    uint8_t *data;
+
+    AIM_ASSERT(cxn->sd == -1);
+
+    /* Increment the generation ID for this connection */
+    cxn->cxn_id += (1 << CXN_ID_GENERATION_SHIFT);
+    cxn->cxn_id &= 0x7fffffff; /* force non-negative */
+
+    LOG_VERBOSE(cxn, "Closing connection with %d bytes in read buf",
+                cxn->read_bytes);
+    cxn->read_bytes = 0;
+    /* Clear write queue */
+    while ((data = bigring_shift(cxn->write_queue))) {
+        LOG_TRACE(cxn, "Freeing outgoing msg %p", data);
+        aim_free(data);
+    }
+
+    cxn->bytes_enqueued = 0;
+    cxn->pkts_enqueued = 0;
+    cxn->write_queue_head_offset = 0;
+
+    cxn->active = false;
+}
+
+
+/**
+ * Attempt to connect() to the specified controller.
+ * Returns INDIGO_ERROR_PARAM or INDIGO_ERROR_UNKNOWN on error,
+ * INDIGO_ERROR_NONE on success,
+ * INDIGO_ERROR_PENDING on in-flight connection.
+ * Caller's responsibility to clean up on error.
+ */
+static indigo_error_t
+cxn_try_to_connect(connection_t *cxn)
+{
+    int rv;
+    struct sockaddr_storage sockaddr;
+
+    if (ind_cxn_parse_sockaddr(&cxn->controller->protocol_params, &sockaddr)
+        != INDIGO_ERROR_NONE) {
+        LOG_INTERNAL(cxn, "Failed to parse address");
+        return INDIGO_ERROR_PARAM;
+    }
+
+    rv = connect(cxn->sd, (struct sockaddr *) &sockaddr, sizeof(sockaddr));
+    if ((rv == 0) ||
+        ((rv == -1) && errno == EISCONN)) {
+        return INDIGO_ERROR_NONE;
+    } else if ((rv == -1) &&
+               (errno == EAGAIN || errno == EALREADY || 
+                errno == EINPROGRESS)) {
+        /* mark socket for writing so that second connect() call can be made,
+         * because nonblocking socket may return EINPROGRESS */
+        ind_soc_data_out_ready(cxn->sd);
+        return INDIGO_ERROR_PENDING;
+    } else {
+        LOG_INTERNAL(cxn, "connect: %s", strerror(errno));
+        if (cxn->aux_id == 0) {
+            cxn->controller->fail_count++;
+        }
+        return INDIGO_ERROR_UNKNOWN;
+    }
+}
+
+
+/**
+ * Start the specified connection.
+ */
+void
+ind_cxn_start(connection_t *cxn)
+{
+    indigo_error_t rv;
+
+    LOG_TRACE(cxn, "Starting cxn %p", cxn);
+
+    AIM_ASSERT(cxn->state == CXN_S_INIT);
+
+    /* register with socket manager to catch socket-writeable event,
+     * so that connect() can successfully complete;
+     * see comment in cxn_try_to_connect():EINPROGRESS handling */
+    ind_soc_socket_register_with_priority(cxn->sd, cxn_socket_ready,
```

Comment:
This call can fail if the fd number is too high. Probably ok to AIM_DIE on failure but we shouldn't ignore it.